diff --git a/source/wifi/Makefile.am b/source/wifi/Makefile.am
index d528d28..4247f52 100644
--- a/source/wifi/Makefile.am
+++ b/source/wifi/Makefile.am
@@ -26,5 +26,5 @@ hardware_platform = i686-linux-gnu
 
 lib_LTLIBRARIES = libhal_wifi.la
 libhal_wifi_la_CPPFLAGS = -I$(top_srcdir)/./include $(CPPFLAGS)
-libhal_wifi_la_SOURCES = wifi_hal.c
+libhal_wifi_la_SOURCES = wifi_hal.c hostapd_cmd.c
 libhal_wifi_la_LDFLAGS = 
